packages/rich-list-core/src/RichListCore.ts

```diff
@@ -0,0 +1,232 @@
+    const tokens = await this._listTokenIds()
+    let updatedBalances = await this._getActiveAddressBalances(tokens, queuedAddressLimit)
+
+    while (Object.keys(updatedBalances).length > 0) {
+      for (const address of Object.keys(updatedBalances)) {
+        const accountAmount = updatedBalances[address]
+        for (const tokenId of tokens) {
+          const balance = accountAmount[tokenId].toNumber()
+          const satoshi = accountAmount[tokenId].times('1e8').dividedToIntegerBy(1).toNumber()
+          await this.addressBalances.put({
+            partition: `${tokenId}`,
+            sort: satoshi,
+            id: `${address}-${tokenId}`,
+            data: {
+              address: address,
+              amount: balance
+            }
+          })
+        }
+      }
+
+      updatedBalances = await this._getActiveAddressBalances(tokens, queuedAddressLimit)
+    }
+  }
+
+  private async _getActiveAddressBalances (tokens: number[], queuedAddressLimit: number): Promise<ActiveAddressAccountAmount> {
+    const queue = await this._addressQueue()
+    const addresses = await queue.receive(queuedAddressLimit)
+
+    const balances: Record<string, AccountAmount> = {}
+    for (const a of addresses) {
+      const nonZeroBalances = await this.apiClient.account.getTokenBalances(
+        { limit: Number.MAX_SAFE_INTEGER },
+        true
+      ) as any as AccountAmount
+      balances[a] = this._appendZeroBalances(nonZeroBalances, tokens)
+      // TBD: should be combine utxo and DFI rich list
+      balances[a]['-1'] = new BigNumber(await this.whaleApiClient.address.getBalance(a))
+    }
+    return balances
+  }
+
+  private _appendZeroBalances (tokenBalances: AccountAmount, tokens: number[]): AccountAmount {
+    const result: AccountAmount = {}
+    for (const t of tokens) {
+      result[t] = tokenBalances[t] ?? new BigNumber(0)
+    }
+    return result
+  }
+
+  private async _getCrawledTip (): Promise<Schema<CrawledBlock> | undefined> {
+    const [lastBlock] = await this.crawledBlockHashes.list({
+      partition: 'NONE',
+      order: 'DESC',
+      limit: 1
+    })
+    return lastBlock
+  }
+
+  private async _invalidateDroppedOutRichList (tokenId: string, height: number): Promise<void> {
+    const lastBlockRichList = await this._getDroppedOutRichList(tokenId, height)
+    await Promise.all(
+      lastBlockRichList.map(async rl => await this.droppedFromRichList.delete(rl.id))
+    )
+  }
+
+  private async _getDroppedOutRichList (tokenId: string, height: number): Promise<Array<Schema<string>>> {
+    return await this.droppedFromRichList.list({
+      limit: Number.MAX_SAFE_INTEGER,
+      partition: tokenId,
+      order: 'DESC',
+      gt: height - 1,
+      lt: height + 1
+    })
+  }
+
+  private async _listTokenIds (): Promise<number[]> {
+    const tokens = await this.apiClient.token.listTokens()
+    return Object.keys(tokens).map(id => Number(id)).concat([-1])
```

Comment:
Regarding this, why are we concatenating -1? I don't quite understand this implementation. What does a token id of -1 indicate?

Reply:
I am using `-1` to represent utxo now, depend if we want to combine it with DFI `0`